theano/tensor/basic.py

```diff
@@ -4606,8 +4612,71 @@ def c_code(self, node, name, inputs, outputs, sub):
         """ % locals()
 
 
+def is_flat(node, outdim=1):
+    """
+    Verifies the dimensionality of the node's variable is equal to
+    outdim. This method is usually called after flatten method on a
+    variable, where the first outdim-1 dimension size(s) of the variable
+    is kept intact, and the last dimension size of the variable is made
+    equal to the multiplication of its remaining dimension size(s), such that
+    the variable would end up with as many dimension as outdim.
+
+    Parameters
+    ----------
+        node : theano.tensor.var.TensorVariable
+            the theano node on which the dimensionality is checked.
+
+        outdim : int
+            the expected dimensionality of node.
+
+    Returns
+    -------
+    bool
+        the comparison result of node's dim
+        and the expected outdim.
+    """
+    return node.ndim == outdim
+
+
 def flatten(x, outdim=1):
-    return Flatten(outdim)(x)
+    """
+    Reshapes the variable x by keeping
+    the first outdim-1 dimension size(s) of x the same,
+    and making the last dimension size of x equal to
+    the multiplication of its remaining dimension size(s).
+
+    Parameters
+    ----------
+        x : theano.tensor.var.TensorVariable
+            the variable that should be reshaped.
+
+        outdim : int
+            the number of dimensions of the returned variable
+
+    Returns
+    -------
+    theano.tensor.var.TensorVariable
+        the flattend variable with dimensionality of outdim
+    """
+    # Any input variable can be flattened to have outdim of 1,
+    # even if it's a scalar. Otherwise, outdim must be positive
+    # and smaller than x.ndim.
+    if outdim < 1 or (outdim > 1 and outdim > x.ndim):
+        raise ValueError('outdim %s out of bound [1, %d)'
+                         % (outdim, x.ndim + 1))
+
+    if outdim > 1:
+        dims = tuple(x.shape[:outdim - 1]) + (-1,)
+    else:
+        dims = (-1,)
+    x_reshaped = x.reshape(dims)
+    bcast_kept_dims = x.broadcastable[:outdim - 1]
+    bcast_new_dim = python_all(x.broadcastable[outdim - 1:])
+    broadcastable = bcast_kept_dims + (bcast_new_dim,)
+    for dim, br in enumerate(broadcastable):
+        if br:
+            x_reshaped = theano.tensor.addbroadcast(x_reshaped, dim)
```

Comment:
I checked the doc and it is file. *args mean that it take a variable number
of arguments, not that it take a list/tuple.

On Thu, Nov 12, 2015 at 8:00 PM, Sina Honari <notifications@github.com>
wrote:

> In theano/tensor/basic.py
> <https://github.com/Theano/Theano/pull/3480#discussion_r44738723>:
>
> > +    # and smaller than x.ndim.
> > +    if outdim < 1 or (outdim > 1 and outdim > x.ndim):
> > +        raise ValueError('outdim %s out of bound [1, %d)'
> > +                         % (outdim, x.ndim + 1))
> > +
> > +    if outdim > 1:
> > +        dims = tuple(x.shape[:outdim - 1]) + (-1,)
> > +    else:
> > +        dims = (-1,)
> > +    x_reshaped = x.reshape(dims)
> > +    bcast_kept_dims = x.broadcastable[:outdim - 1]
> > +    bcast_new_dim = python_all(x.broadcastable[outdim - 1:])
> > +    broadcastable = bcast_kept_dims + (bcast_new_dim,)
> > +    for dim, br in enumerate(broadcastable):
> > +        if br:
> > +            x_reshaped = theano.tensor.addbroadcast(x_reshaped, dim)
>
> You correction works, however, the documentation on Theano website (
> http://deeplearning.net/software/theano/library/tensor/basic.html) seems
> wrong, since just passing a list or tuple throws this error:
>
> list error:
> x = T.tensor4()
> T.addbroadcast(x, [0,1,2])
> TypeError: unhashable type: 'list'
>
> tuple error:
> T.addbroadcast(x, (0,1,2))
> AssertionError: ('Rebroadcast needs integer axes. Got ', (0, 1, 2))
>
> The star(*) that unpacks list or tuple should also be passed to make it
> work
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3480/files#r44738723>.
>